File: 3620-maximum-number-of-distinct-elements-after-operations/maximum-number-of-distinct-elements-after-operations.java
class Solution {
    public int maxDistinctElements(int[] nums, int k) {
        Arrays.sort(nums);
        int prevMax = Integer.MIN_VALUE;
        int distinct = 0;

        for (int num : nums) {
            int lowerBound = num - k;
            int upperBound = num + k;

            if (prevMax < lowerBound) {
                prevMax = lowerBound;
                distinct++;
            } else if (prevMax < upperBound) {
                prevMax++;
                distinct++;
            }
        }

        return distinct;
    }
}

File: 3620-maximum-number-of-distinct-elements-after-operations/maximum-number-of-distinct-elements-after-operations.go
func maxDistinctElements(nums []int, k int) int {
	sort.Ints(nums)
	last := nums[0] - k - 1
	count := 0

	for _, num := range nums {
		if num-k > last {
			last = num - k
		} else if num+k <= last {
			continue
		} else {
			last++
		}
		count++
	}

	return count
}